### ISSUE TITLE
compatibility mode not accurately emulating IE. 

### ISSUE THREAD

**Issue by lakshay35** [open-wc/open-wc#1403]
I changed our local es-dev-server to use build files from rollup and am getting errors with the legacy version loading in IE 11. I'd suggest we change es-dev-server to act like angular's local dev server that compiles source files and outputs everything.

Aditionally, I have an error with my build output where the following error appears in IE when I import the component.

```
Unhandled promise rejection TypeError: @@toPrimitive must return a primitive value.
```

I am bundling chart.js which is just an import inside my web component. The chartjs import is a bundled file from another rollup output by the author. Anybody know how I can ignore my rollup configuration from ignoring and rollup transformations on that and just adding it to the output?

Following is my rollup config:

```
// import createDefaultConfig from '@open-wc/building-rollup/modern-and-legacy-config';
import { createCompatibilityConfig } from '@open-wc/building-rollup';
import replace from 'rollup-plugin-replace';
import typescript from '@rollup/plugin-typescript';
import babel from 'rollup-plugin-babel';
// if you need to support IE11 use 'modern-and-legacy-config' instead.
// import createDefaultConfig from '@open-wc/building-rollup/modern-and-legacy-config';

const configs = createCompatibilityConfig({
  production: true,
  input: './index.html',
  outputDir: 'dist-dev',
  plugins: [
    replace({
      'process.env.NODE_ENV': 'production',
    }),
    babel({}),
    typescript(),
  ],
  extensions: ['.js', '.ts', '.mjs'],
});

export default configs.map(config => ({
  ...config,
  output: {
    ...config.output,
    entryFileNames: '[name].js',
    chunkFileNames: '[name].js',
  },
}));

```

**Comment by LarsDenBakker** [comment]
I'm not quite sure about your questions, it's like three questions in one? :)

I don't know how the angular dev server works, so I don't know what you're asking us to match. 

It sounds like you're running into some issue with polyfills. Can you try running with `--compatibility none`? There is no need to run compatibility mode on an already built application.

For general rollup support it's best to check with the rollup team, they will be better

**Comment by frank-dspeed** [comment]
@LarsDenBakker do not refer that to rollup its a bundler we can not help with pollyfills

**Comment by lakshay35** [comment]
How does the es-dev-server apply polyfills? It seems like its not doing what systemjs in production is doing.

**Comment by LarsDenBakker** [comment]
They are added to the index.html using https://github.com/open-wc/open-wc/tree/master/packages/polyfills-loader

**Comment by lakshay35** [comment]
I was looking through the versions of the polyfills the index-html plugin is leveraging and the ones that the polyfills loader is referencing. Any idea what might be causing the difference in behavior? 

I am currently solving the problem of real-world browser emulation by running rollup in watch mode with the es-dev-server, which slows us down due to the lag with rollup and we are unable to use chrome debugging tools cause the code is minified.

**Comment by LarsDenBakker** [comment]
It's hard to say without a clearer explanation of the issue, what setup do you have for building and development, what goes wrong exactly? A minimal reproduction would help.

**Comment by lakshay35** [comment]
so I have a web component that imports imports the chart.js build file like so:

```
import 'chart.js'
```

The component renders a simple chart. I use compatibility auto for the es-dev-server and it woks in each browser including IE 11 in my local dev environment. But when I build the component using open-wc's configuration it throws an error in the IE 11 browser console.

I'm only trying to solve that inconsistency. Some of the styles in the browser that work locally also get messed up in IE when built.

**Comment by LarsDenBakker** [comment]
`chart.js` refers to this package https://www.npmjs.com/package/chartjs ?

**Comment by LarsDenBakker** [comment]
I tried a clean project with `import 'chart.js'` but I didn't get any errors during development or build.

**Comment by lakshay35** [comment]
Did you use https://www.npmjs.com/package/chart.js? There are no errors during development. There is a warning during build that 'this has been rewritten to undefined'. Then the build file for the component import chart.js and rendering a chart, does not render in IE 11 only. It throws an error saying '@@toPrimitive ...' it only happens when I import that js file into a typescript component. In javascript it works fine.

**Comment by lakshay35** [comment]
Additionally I can get the component to render in IE when I change my babelrc from

```
{
  "presets": ["@babel/preset-typescript"],
  "plugins": [
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true }],
    "@babel/plugin-proposal-class-properties"
  ]
}
```

to 

```
{
  "presets": ["@babel/preset-typescript"],
  "plugins": [
    "@babel/plugin-proposal-decorators", { "legacy": true }],
  ["@babel/plugin-proposal-class-properties", { "loose": true }]
}
```

but then I can't get the component lifecyle to update anymore

**Comment by LarsDenBakker** [comment]
I created a clean project with `npm init @open-wc` and used `import 'chart.js'` as you mentioned, and I don't get an error with es-dev-server or rollup.

Have you tried running the built output with `--compatibility none` ?

**Comment by lakshay35** [comment]
yes I did and it returns that error (i.e. @@toPrimitive...)

**Comment by lakshay35** [comment]
What was your babelrc file and the component? Can you post it here?

**Comment by LarsDenBakker** [comment]
I did a clean `npm init @open-wc` and added the import as you explained. 

For further triage we need to have a minimal reproduction. If it's related to babel plugins, I'm not sure if it's related to open-wc specific setup?

**Comment by emilio-martinez** [comment]
TL;DR: This is a polyfill + babel issue, not an open-wc issue.

That error is related to the `_toPrimitive` helper that Babel emits, which will only be emitted for the "spec" (i.e., not legacy) decorators when using LitElement. That helper function will fail a `typeof` check against Symbols without applying `@babel/plugin-transform-typeof-symbol` or similar. Without that,  the function will proceed with other checks that result in the `@@toPrimitive` error you're seeing.

I don't believe open-wc is excluding that plugin anywhere (whereas create-react-app does, for example), so this is definitely not an issue for this repo.

**Comment by LarsDenBakker** [comment]
Thanks a lot for that clarification

**Comment by LarsDenBakker** [comment]
Going to close this one as there isn't an actionable item for us here.
